File: apps/quickink/android/app/src/main/java/app/quickink/mobile/features/stories/StoryImageExporter.kt
/*
 * StoryImageExporter.kt
 *
 * Stories Phase 4 — composites a story into a single tall PNG.
 * Mirror of iOS `StoryImageExporter.swift`; capped at 12 items, with
 * a "+ N more in the app" footer when truncated.
 */

package app.quickink.mobile.features.stories

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.LinearGradient
import android.graphics.Paint
import android.graphics.Path
import android.graphics.RectF
import android.graphics.Shader
import android.net.Uri
import android.text.Layout
import android.text.StaticLayout
import android.text.TextPaint
import app.quickink.mobile.data.story.StoryEntity
import app.quickink.mobile.data.storyitem.StoryItemEntity
import java.io.File
import java.io.FileOutputStream

object StoryImageExporter {

    private const val MAX_ITEMS = 12
    private const val WIDTH     = 1080f
    private const val PAD       = 56f

    fun export(
        context: Context,
        story: StoryEntity,
        items: List<StoryItemEntity>,
        previewUris: Map<String, String> = emptyMap(),
    ): File {
        val limited = items.take(MAX_ITEMS)
        val overflow = (items.size - limited.size).coerceAtLeast(0)
        val height   = computeHeight(limited, overflow > 0)

        val bitmap = Bitmap.createBitmap(WIDTH.toInt(), height.toInt(), Bitmap.Config.ARGB_8888)
        val canvas = Canvas(bitmap)
        draw(canvas, story, limited, overflow, height, previewUris)

        val dir = File(context.cacheDir, "stories-exports").apply { if (!exists()) mkdirs() }
        val file = File(dir, "${sanitizeFileName(story.title)}.png")
        FileOutputStream(file).use { out ->
            bitmap.compress(Bitmap.CompressFormat.PNG, 100, out)
        }
        bitmap.recycle()
        return file
    }

    private fun computeHeight(items: List<StoryItemEntity>, hasOverflow: Boolean): Float {
        var total = 560f + 60f
        items.forEach { total += rowHeight(it) }
        if (hasOverflow) total += 56f
        total += 200f
        return total
    }

    private fun rowHeight(item: StoryItemEntity): Float = when (item.kind) {
        StoryItemEntity.Kind.TEXT_BLOCK.raw        -> 120f
        StoryItemEntity.Kind.HANDWRITTEN_NOTE.raw  -> 96f
        StoryItemEntity.Kind.DATE_DIVIDER.raw      -> 56f
        StoryItemEntity.Kind.PLACE_PIN.raw         -> 48f
        StoryItemEntity.Kind.VOICE_CLIP.raw        -> 96f
        else -> when (item.layout) {
            StoryItemEntity.Layout.HALF.raw -> 320f
            StoryItemEntity.Layout.GRID.raw -> 240f
            else                             -> 520f
        }
    }

    private fun draw(canvas: Canvas, story: StoryEntity, items: List<StoryItemEntity>, overflow: Int, height: Float, previewUris: Map<String, String>) {
        // Cream background
        canvas.drawColor(Color.parseColor("#FBF6EE"))

        // Cover
        val coverRect = RectF(PAD, PAD, WIDTH - PAD, PAD + 480f)
        val (startCol, endCol) = coverColors(story.coverStyle)
        val coverPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            shader = LinearGradient(
                coverRect.left, coverRect.top, coverRect.right, coverRect.bottom,
                startCol, endCol, Shader.TileMode.CLAMP,
            )
        }
        canvas.drawRoundRect(coverRect, 24f, 24f, coverPaint)

        // Title at bottom-left
        val titlePaint = TextPaint(Paint.ANTI_ALIAS_FLAG).apply {
            color = Color.BLACK
            textSize = 60f
            typeface = android.graphics.Typeface.create(android.graphics.Typeface.SERIF, android.graphics.Typeface.NORMAL)
        }
        canvas.drawText(story.title, coverRect.left + 32f, coverRect.bottom - 100f, titlePaint)
        if (!story.subtitle.isNullOrEmpty()) {
            val subPaint = TextPaint(Paint.ANTI_ALIAS_FLAG).apply {
                color = Color.parseColor("#333333")
                textSize = 36f
                typeface = android.graphics.Typeface.create(android.graphics.Typeface.SERIF, android.graphics.Typeface.ITALIC)
            }
            canvas.drawText(story.subtitle, coverRect.left + 32f, coverRect.bottom - 40f, subPaint)
        }

        // Attribution
        var y = coverRect.bottom + 40f
        val attrPaint = TextPaint(Paint.ANTI_ALIAS_FLAG).apply {
            color = Color.DKGRAY
            textSize = 22f
            typeface = android.graphics.Typeface.create(android.graphics.Typeface.SERIF, android.graphics.Typeface.ITALIC)
            textAlign = Paint.Align.CENTER
        }
        canvas.drawText("${items.size} items", WIDTH / 2f, y, attrPaint)
        y += 50f

        // Items + markers
        val markers = StoryDayMarkers.derive(items).associateBy { it.precedingItemId }
        for (item in items) {
            markers[item.id]?.let { marker ->
                val mp = TextPaint(Paint.ANTI_ALIAS_FLAG).apply {
                    color = Color.parseColor("#C65A3E")
                    textSize = 20f
                    typeface = android.graphics.Typeface.create(android.graphics.Typeface.SERIF, android.graphics.Typeface.ITALIC)
                    letterSpacing = 0.05f
                }
                canvas.drawText(marker.label, PAD, y + 16f, mp)
                y += 40f
            }
            val rect = RectF(PAD, y, WIDTH - PAD, y + rowHeight(item))
            drawRow(canvas, item, rect, previewUris)
            y = rect.bottom + 16f
        }

        if (overflow > 0) {
            val font = TextPaint(Paint.ANTI_ALIAS_FLAG).apply {
                color = Color.GRAY
                textSize = 20f
                typeface = android.graphics.Typeface.create(android.graphics.Typeface.SERIF, android.graphics.Typeface.ITALIC)
                textAlign = Paint.Align.CENTER
            }
            canvas.drawText("+ $overflow more in the app", WIDTH / 2f, y + 24f, font)
            y += 56f
        }

        // End-card + footer
        y += 32f
        val endRect = RectF(PAD, y, WIDTH - PAD, y + 96f)
        val stroke = Paint(Paint.ANTI_ALIAS_FLAG).apply {
            style = Paint.Style.STROKE; strokeWidth = 1f; color = Color.LTGRAY
        }
        canvas.drawRoundRect(endRect, 24f, 24f, stroke)
        val endPaint = TextPaint(Paint.ANTI_ALIAS_FLAG).apply {
            color = Color.DKGRAY
            textSize = 22f
            letterSpacing = 0.15f
            typeface = android.graphics.Typeface.create(android.graphics.Typeface.SERIF, android.graphics.Typeface.ITALIC)
            textAlign = Paint.Align.CENTER
        }
        canvas.drawText("— THE END —", WIDTH / 2f, endRect.centerY() + 8f, endPaint)
        y = endRect.bottom + 24f
        val footerPaint = TextPaint(Paint.ANTI_ALIAS_FLAG).apply {
            color = Color.GRAY
            textSize = 18f
            typeface = android.graphics.Typeface.create(android.graphics.Typeface.SERIF, android.graphics.Typeface.ITALIC)
            textAlign = Paint.Align.CENTER
        }
        canvas.drawText("— Made with QuickInk —", WIDTH / 2f, y + 8f, footerPaint)
    }

    private fun drawRow(canvas: Canvas, item: StoryItemEntity, rect: RectF, previewUris: Map<String, String> = emptyMap()) {
        when (item.kind) {
            StoryItemEntity.Kind.TEXT_BLOCK.raw -> {
                drawWrappedText(canvas, item.text.orEmpty(), rect, 22f, italic = false, color = Color.BLACK)
            }
            StoryItemEntity.Kind.HANDWRITTEN_NOTE.raw -> {
                val rule = Paint().apply { color = Color.parseColor("#C65A3E") }
                canvas.drawRect(rect.left, rect.top, rect.left + 4f, rect.bottom, rule)
                drawWrappedText(
                    canvas, item.text.orEmpty(),
                    RectF(rect.left + 24f, rect.top + 8f, rect.right, rect.bottom),
                    28f, italic = true, color = Color.DKGRAY,
                )
            }
            StoryItemEntity.Kind.DATE_DIVIDER.raw -> {
                val paint = TextPaint(Paint.ANTI_ALIAS_FLAG).apply {
                    color = Color.parseColor("#C65A3E")
                    textSize = 22f
                    typeface = android.graphics.Typeface.create(android.graphics.Typeface.SERIF, android.graphics.Typeface.ITALIC)
                    letterSpacing = 0.12f
                }
                canvas.drawText(item.text ?: "Date divider", rect.left, rect.top + 24f, paint)
            }
            StoryItemEntity.Kind.PLACE_PIN.raw -> {
                val paint = TextPaint(Paint.ANTI_ALIAS_FLAG).apply {
                    color = Color.DKGRAY
                    textSize = 22f
                    typeface = android.graphics.Typeface.create(android.graphics.Typeface.SERIF, android.graphics.Typeface.ITALIC)
                }
                canvas.drawText("📍 ${item.text.orEmpty()}", rect.left, rect.top + 24f, paint)
            }
            StoryItemEntity.Kind.VOICE_CLIP.raw -> {
                val stroke = Paint(Paint.ANTI_ALIAS_FLAG).apply {
                    style = Paint.Style.STROKE; strokeWidth = 1f; color = Color.LTGRAY
                }
                canvas.drawRoundRect(rect, 16f, 16f, stroke)
                val paint = TextPaint(Paint.ANTI_ALIAS_FLAG).apply {
                    color = Color.DKGRAY
                    textSize = 20f
                    typeface = android.graphics.Typeface.create(android.graphics.Typeface.SERIF, android.graphics.Typeface.ITALIC)
                }
                canvas.drawText("🎙  ${item.caption ?: "Voice clip"}",
                    rect.left + 24f, rect.centerY() + 6f, paint)
            }
            else -> {
                val photoHeight = (rect.height() - 60f).coerceAtLeast(120f)
                val photoRect = RectF(rect.left, rect.top, rect.right, rect.top + photoHeight)
                val bmp = item.refId?.let { previewUris[it] }?.let { loadBitmap(it) }
                if (bmp != null) {
                    canvas.save()
                    val clip = Path().apply { addRoundRect(photoRect, 16f, 16f, Path.Direction.CW) }
                    canvas.clipPath(clip)
                    val dst = aspectFill(bmp.width, bmp.height, photoRect)
                    canvas.drawBitmap(bmp, null, dst, null)
                    canvas.restore()
                    bmp.recycle()
                } else {
                    val fill = Paint(Paint.ANTI_ALIAS_FLAG).apply { color = Color.parseColor("#EAE5DA") }
                    canvas.drawRoundRect(photoRect, 16f, 16f, fill)
                }
                if (!item.caption.isNullOrEmpty()) {
                    val paint = TextPaint(Paint.ANTI_ALIAS_FLAG).apply {
                        color = Color.DKGRAY
                        textSize = 20f
                        typeface = android.graphics.Typeface.create(android.graphics.Typeface.SERIF, android.graphics.Typeface.ITALIC)
                    }
                    canvas.drawText(item.caption, rect.left, photoRect.bottom + 24f, paint)
                }
            }
        }
    }

    private fun loadBitmap(uri: String): Bitmap? = runCatching {
        val parsed = if (uri.startsWith("file://")) Uri.parse(uri) else null
        val path = parsed?.path ?: uri
        BitmapFactory.decodeFile(path)
    }.getOrNull()

    private fun aspectFill(w: Int, h: Int, target: RectF): RectF {
        if (w <= 0 || h <= 0) return target
        val scale = maxOf(target.width() / w, target.height() / h)
        val outW = w * scale
        val outH = h * scale
        val cx = (target.left + target.right) / 2f
        val cy = (target.top + target.bottom) / 2f
        return RectF(cx - outW / 2f, cy - outH / 2f, cx + outW / 2f, cy + outH / 2f)
    }

    private fun drawWrappedText(canvas: Canvas, text: String, rect: RectF, size: Float, italic: Boolean, color: Int) {
        val paint = TextPaint(Paint.ANTI_ALIAS_FLAG).apply {
            this.color = color
            this.textSize = size
            typeface = android.graphics.Typeface.create(
                android.graphics.Typeface.SERIF,
                if (italic) android.graphics.Typeface.ITALIC else android.graphics.Typeface.NORMAL,
            )
        }
        val width = rect.width().toInt().coerceAtLeast(1)
        val layout = StaticLayout.Builder
            .obtain(text, 0, text.length, paint, width)
            .setAlignment(Layout.Alignment.ALIGN_NORMAL)
            .setLineSpacing(6f, 1f)
            .setIncludePad(false)
            .build()
        canvas.save()
        canvas.translate(rect.left, rect.top)
        layout.draw(canvas)
        canvas.restore()
    }

    private fun coverColors(coverStyleRaw: String): Pair<Int, Int> = when (coverStyleRaw) {
        StoryEntity.CoverStyle.GRADIENT.raw    -> Color.parseColor("#E07856") to Color.parseColor("#C65A3E")
        StoryEntity.CoverStyle.TYPOGRAPHIC.raw -> Color.parseColor("#EADFCF") to Color.parseColor("#E8DCC4")
        else                                   -> Color.parseColor("#E8DCC4") to Color.parseColor("#F0E4D7")
    }

    private fun sanitizeFileName(name: String): String {
        val out = name.map { c -> if (c in "/\\:*?\"<>|") '-' else c }.joinToString("").trim()
        return out.ifBlank { "Story" }
    }
}

File: apps/quickink/ios/QuickInk/Stories/StoryImageExporter.swift
/*
 * StoryImageExporter.swift
 *
 * Stories Phase 4 — composites a Story + items into a single tall
 * PNG ("a tall PNG for chats or stories", §7.5 mockup). Capped at 12
 * items per image; if a story has more, the export truncates and the
 * last visible row gets a "+ N more in the app" footer line.
 *
 * Rendering goes through `UIGraphicsImageRenderer` at @2x for retina
 * sharpness. The output is byte-stable for the same input.
 *
 * Mirror of Android `StoryImageExporter.kt`.
 */

import Foundation
import UIKit

enum StoryImageExporter {

    /// Item cap per the handoff doc.
    private static let maxItems = 12
    /// Canvas width (1080 px is the standard "story" aspect).
    private static let width: CGFloat  = 1080
    /// Padding around the page content.
    private static let pad: CGFloat    = 56

    /// Render to a temp PNG on disk; returns the file URL.
    /// See `StoryPdfExporter.export` for the `previewUris` contract.
    static func export(story: Story, items: [StoryItem], previewUris: [String: String] = [:]) throws -> URL {
        let limited = Array(items.prefix(maxItems))
        let overflow = max(0, items.count - limited.count)
        let height = computeHeight(items: limited, hasOverflow: overflow > 0)

        let renderer = UIGraphicsImageRenderer(
            size: CGSize(width: width, height: height),
            format: { let f = UIGraphicsImageRendererFormat(); f.scale = 2; return f }()
        )
        let image = renderer.image { rendererCtx in
            draw(story: story, items: limited, overflow: overflow, previewUris: previewUris, in: CGRect(x: 0, y: 0, width: width, height: height))
        }
        guard let data = image.pngData() else {
            throw NSError(domain: "StoryImageExporter", code: 1)
        }
        let url = FileManager.default.temporaryDirectory
            .appendingPathComponent("\(sanitizeFileName(story.title)).png")
        try data.write(to: url, options: .atomic)
        return url
    }

    // MARK: - Layout

    private static func computeHeight(items: [StoryItem], hasOverflow: Bool) -> CGFloat {
        // Rough vertical budget: cover (560) + per-item heights + end-card.
        var total: CGFloat = 560 + 60 // cover + attribution
        for item in items {
            total += rowHeight(for: item)
        }
        if hasOverflow { total += 56 }
        total += 200 // end-card
        return total
    }

    private static func rowHeight(for item: StoryItem) -> CGFloat {
        switch item.kind {
        case StoryItem.Kind.textBlock.rawValue:        return 120
        case StoryItem.Kind.handwrittenNote.rawValue:  return 96
        case StoryItem.Kind.dateDivider.rawValue:      return 56
        case StoryItem.Kind.placePin.rawValue:         return 48
        case StoryItem.Kind.voiceClip.rawValue:        return 96
        default:
            switch item.layout {
            case StoryItem.Layout.half.rawValue: return 320
            case StoryItem.Layout.grid.rawValue: return 240
            default:                             return 520
            }
        }
    }

    // MARK: - Drawing

    private static func draw(story: Story, items: [StoryItem], overflow: Int, previewUris: [String: String], in rect: CGRect) {
        guard let ctx = UIGraphicsGetCurrentContext() else { return }

        // ---- Canvas background (warm cream) ----
        ctx.setFillColor(UIColor(red: 0xFB/255, green: 0xF6/255, blue: 0xEE/255, alpha: 1).cgColor)
        ctx.fill(rect)

        // ---- Cover (560 tall) ----
        let coverRect = CGRect(x: pad, y: pad, width: rect.width - pad * 2, height: 480)
        let colors = coverColors(for: story.coverStyle)
        let gradient = CGGradient(
            colorsSpace: CGColorSpaceCreateDeviceRGB(),
            colors: [colors.0.cgColor, colors.1.cgColor] as CFArray,
            locations: [0, 1]
        )!
        ctx.saveGState()
        ctx.addPath(UIBezierPath(roundedRect: coverRect, cornerRadius: 24).cgPath)
        ctx.clip()
        ctx.drawLinearGradient(gradient,
            start: CGPoint(x: coverRect.minX, y: coverRect.minY),
            end:   CGPoint(x: coverRect.maxX, y: coverRect.maxY),
            options: [])
        ctx.restoreGState()

        // Title + subtitle bottom-left
        let titleFont = UIFont(name: "Lora-Medium", size: 60) ?? UIFont.systemFont(ofSize: 60, weight: .medium)
        let titleAttrs: [NSAttributedString.Key: Any] = [
            .font: titleFont, .foregroundColor: UIColor.black,
        ]
        let titleRect = CGRect(
            x: coverRect.minX + 32,
            y: coverRect.maxY - 160,
            width: coverRect.width - 64,
            height: 120
        )
        (story.title as NSString).draw(
            with: titleRect,
            options: [.usesLineFragmentOrigin],
            attributes: titleAttrs,
            context: nil
        )
        if let subtitle = story.subtitle, !subtitle.isEmpty {
            let subFont = UIFont(name: "Caveat-Medium", size: 36) ?? UIFont.italicSystemFont(ofSize: 32)
            (subtitle as NSString).draw(
                at: CGPoint(x: coverRect.minX + 32, y: coverRect.maxY - 64),
                withAttributes: [.font: subFont, .foregroundColor: UIColor(white: 0.2, alpha: 1)]
            )
        }

        // ---- Attribution ----
        var y: CGFloat = coverRect.maxY + 40
        let attrFont = UIFont(name: "Lora-Italic", size: 22) ?? UIFont.italicSystemFont(ofSize: 22)
        let attribution = "\(items.count) items"
        let attrAttrs: [NSAttributedString.Key: Any] = [.font: attrFont, .foregroundColor: UIColor.darkGray]
        let attribSize = (attribution as NSString).size(withAttributes: attrAttrs)
        (attribution as NSString).draw(
            at: CGPoint(x: (rect.width - attribSize.width) / 2, y: y),
            withAttributes: attrAttrs
        )
        y += attribSize.height + 32

        // ---- Items + sticky day markers ----
        let markers = StoryDayMarkers.derive(from: items)
        let markerByItemId = Dictionary(uniqueKeysWithValues: markers.map { ($0.precedingItemId, $0) })
        let contentLeft  = pad
        let contentRight = rect.width - pad
        let contentWidth = contentRight - contentLeft

        for item in items {
            if let marker = markerByItemId[item.id] {
                let mfont = UIFont(name: "Lora-Italic", size: 20) ?? UIFont.italicSystemFont(ofSize: 20)
                (marker.label as NSString).draw(
                    at: CGPoint(x: contentLeft, y: y),
                    withAttributes: [
                        .font: mfont,
                        .foregroundColor: UIColor.systemOrange.withAlphaComponent(0.85),
                        .kern: 1,
                    ]
                )
                y += 36
            }
            let itemRect = CGRect(x: contentLeft, y: y, width: contentWidth, height: rowHeight(for: item))
            drawRow(item, in: itemRect, previewUris: previewUris)
            y += itemRect.height + 16
        }

        if overflow > 0 {
            let font = UIFont(name: "Lora-Italic", size: 20) ?? UIFont.italicSystemFont(ofSize: 20)
            let msg = "+ \(overflow) more in the app"
            let size = (msg as NSString).size(withAttributes: [.font: font])
            (msg as NSString).draw(
                at: CGPoint(x: (rect.width - size.width) / 2, y: y),
                withAttributes: [.font: font, .foregroundColor: UIColor.gray]
            )
            y += 48
        }

        // ---- End-card + footer ----
        y += 32
        let endRect = CGRect(x: pad, y: y, width: contentWidth, height: 96)
        ctx.setStrokeColor(UIColor.lightGray.cgColor)
        ctx.setLineWidth(1)
        ctx.addPath(UIBezierPath(roundedRect: endRect, cornerRadius: 24).cgPath)
        ctx.strokePath()
        let endFont = UIFont(name: "Lora-Italic", size: 22) ?? UIFont.italicSystemFont(ofSize: 22)
        let endLabel = "— THE END —"
        let endSize = (endLabel as NSString).size(withAttributes: [.font: endFont, .kern: 2])
        (endLabel as NSString).draw(
            at: CGPoint(x: endRect.midX - endSize.width / 2, y: endRect.midY - 14),
            withAttributes: [.font: endFont, .foregroundColor: UIColor.darkGray, .kern: 2]
        )
        y = endRect.maxY + 24
        let footer = "— Made with QuickInk —"
        let footerFont = UIFont(name: "Lora-Italic", size: 18) ?? UIFont.italicSystemFont(ofSize: 18)
        let footerSize = (footer as NSString).size(withAttributes: [.font: footerFont])
        (footer as NSString).draw(
            at: CGPoint(x: (rect.width - footerSize.width) / 2, y: y),
            withAttributes: [.font: footerFont, .foregroundColor: UIColor.gray]
        )
    }

    private static func drawRow(_ item: StoryItem, in rect: CGRect, previewUris: [String: String] = [:]) {
        guard let ctx = UIGraphicsGetCurrentContext() else { return }
        switch item.kind {
        case StoryItem.Kind.textBlock.rawValue:
            let font = UIFont(name: "Lora-Regular", size: 22) ?? UIFont.systemFont(ofSize: 22)
            let para = NSMutableParagraphStyle(); para.lineSpacing = 6
            (item.text ?? "").draw(
                with: rect, options: [.usesLineFragmentOrigin],
                attributes: [.font: font, .foregroundColor: UIColor.black, .paragraphStyle: para],
                context: nil
            )
        case StoryItem.Kind.handwrittenNote.rawValue:
            ctx.setFillColor(UIColor.systemOrange.cgColor)
            ctx.fill(CGRect(x: rect.minX, y: rect.minY, width: 4, height: rect.height))
            let font = UIFont(name: "Caveat-Medium", size: 28) ?? UIFont.italicSystemFont(ofSize: 28)
            (item.text ?? "").draw(
                with: rect.insetBy(dx: 20, dy: 0).offsetBy(dx: 12, dy: 8),
                options: [.usesLineFragmentOrigin],
                attributes: [.font: font, .foregroundColor: UIColor.darkGray],
                context: nil
            )
        case StoryItem.Kind.dateDivider.rawValue:
            let font = UIFont(name: "Lora-Italic", size: 22) ?? UIFont.italicSystemFont(ofSize: 22)
            let label = item.text ?? "Date divider"
            (label as NSString).draw(
                at: CGPoint(x: rect.minX, y: rect.minY),
                withAttributes: [.font: font, .foregroundColor: UIColor.systemOrange.withAlphaComponent(0.85), .kern: 1.5]
            )
        case StoryItem.Kind.placePin.rawValue:
            let font = UIFont(name: "Lora-MediumItalic", size: 22) ?? UIFont.italicSystemFont(ofSize: 22)
            ("📍 " + (item.text ?? "")).draw(
                at: CGPoint(x: rect.minX, y: rect.minY),
                withAttributes: [.font: font, .foregroundColor: UIColor.darkGray]
            )
        case StoryItem.Kind.voiceClip.rawValue:
            ctx.setStrokeColor(UIColor.lightGray.cgColor); ctx.setLineWidth(1)
            ctx.addPath(UIBezierPath(roundedRect: rect, cornerRadius: 16).cgPath); ctx.strokePath()
            let font = UIFont(name: "Lora-Italic", size: 20) ?? UIFont.italicSystemFont(ofSize: 20)
            ("🎙  " + (item.caption ?? "Voice clip")).draw(
                at: CGPoint(x: rect.minX + 24, y: rect.midY - 12),
                withAttributes: [.font: font, .foregroundColor: UIColor.darkGray]
            )
        default:
            let photoHeight: CGFloat = max(120, rect.height - 60)
            let photoRect = CGRect(x: rect.minX, y: rect.minY, width: rect.width, height: photoHeight)

            let img: UIImage? = {
                guard let refId = item.refId,
                      let uri   = previewUris[refId] else { return nil }
                return loadImage(at: uri)
            }()

            if let img = img {
                ctx.saveGState()
                ctx.addPath(UIBezierPath(roundedRect: photoRect, cornerRadius: 16).cgPath)
                ctx.clip()
                img.draw(in: aspectFill(image: img.size, into: photoRect))
                ctx.restoreGState()
            } else {
                ctx.setFillColor(UIColor(white: 0.92, alpha: 1).cgColor)
                ctx.addPath(UIBezierPath(roundedRect: photoRect, cornerRadius: 16).cgPath)
                ctx.fillPath()
            }

            if let caption = item.caption, !caption.isEmpty {
                let font = UIFont(name: "Lora-Italic", size: 20) ?? UIFont.italicSystemFont(ofSize: 20)
                (caption as NSString).draw(
                    at: CGPoint(x: rect.minX, y: photoRect.maxY + 12),
                    withAttributes: [.font: font, .foregroundColor: UIColor.darkGray]
                )
            }
        }
    }

    private static func aspectFill(image imageSize: CGSize, into target: CGRect) -> CGRect {
        guard imageSize.width > 0, imageSize.height > 0 else { return target }
        let scale = max(target.width / imageSize.width, target.height / imageSize.height)
        let w = imageSize.width  * scale
        let h = imageSize.height * scale
        return CGRect(
            x: target.midX - w / 2,
            y: target.midY - h / 2,
            width:  w,
            height: h
        )
    }

    private static func loadImage(at uri: String) -> UIImage? {
        let path: String?
        if let url = URL(string: uri), url.isFileURL {
            path = url.path
        } else {
            path = uri
        }
        guard let p = path, FileManager.default.fileExists(atPath: p) else { return nil }
        return UIImage(contentsOfFile: p)
    }

    // MARK: - Helpers (mirror StoryPdfExporter)

    private static func coverColors(for coverStyleRaw: String) -> (UIColor, UIColor) {
        switch coverStyleRaw {
        case Story.CoverStyle.gradient.rawValue:
            return (UIColor(red: 0xE0/255, green: 0x78/255, blue: 0x56/255, alpha: 1),
                    UIColor(red: 0xC6/255, green: 0x5A/255, blue: 0x3E/255, alpha: 1))
        case Story.CoverStyle.typographic.rawValue:
            return (UIColor(red: 0xEA/255, green: 0xDF/255, blue: 0xCF/255, alpha: 1),
                    UIColor(red: 0xE8/255, green: 0xDC/255, blue: 0xC4/255, alpha: 1))
        default:
            return (UIColor(red: 0xE8/255, green: 0xDC/255, blue: 0xC4/255, alpha: 1),
                    UIColor(red: 0xF0/255, green: 0xE4/255, blue: 0xD7/255, alpha: 1))
        }
    }

    private static func sanitizeFileName(_ name: String) -> String {
        let invalid = CharacterSet(charactersIn: "/\\:*?\"<>|")
        let cleaned = name.unicodeScalars
            .map { invalid.contains($0) ? "-" : String($0) }
            .joined()
            .trimmingCharacters(in: .whitespaces)
        return cleaned.isEmpty ? "Story" : cleaned
    }
}
